File: activeobject/src/domain_name_registry.rb
class DomainNameRegistry
  def initialize(domain_space)
    @domain_space = domain_space
    @registry = []
  end
  def buy_domain_name(domain_name)
    full_domain_name = "#{domain_name}.#{@domain_space}"
    sleep 2
    if (@registry.include?(domain_name))
      return PurchaseResult.new(full_domain_name, false, "Not Available")
    end
    @registry << domain_name
    return PurchaseResult.new(full_domain_name, true, "Purchased") 
  end
end

class PurchaseResult
  attr_reader :full_domain_name
  attr_reader :success
  attr_reader :message
  def initialize(full_domain_name, success, message)
    @full_domain_name = full_domain_name
    @success = success
    @message = message
  end
  def to_s
    return @full_domain_name + ": " + @message
  end
end

File: trunk/activeobject/java/src/activeobject/DomainNameRegistry.java
package activeobject;

import java.util.HashSet;
import java.util.Set;

public class DomainNameRegistry {

	private final String domainSpace;
	private Set<String> registry = new HashSet<String>();

	public DomainNameRegistry(String domainSpace) {
		this.domainSpace = domainSpace;
	}
	
	public PurchaseResult buyDomainName(String domainName)
	{
		// Pretend remote call takes 2 seconds.
		sleep(2);
		if (registry.contains(domainName)) {
			return new SimplePurchaseResult(fullDomainName(domainName), false, "Not Available");
		}
		registry.add(domainName);
		return new SimplePurchaseResult(fullDomainName(domainName), true, "Purchased");
	}

	private String fullDomainName(String domainName) {
		return domainName + "." + domainSpace;
	}

	private void sleep(int secs) {
		try {
			Thread.sleep(secs * 1000);
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
}
